### ISSUE TITLE
sign tarballs with PGP instead of using a sha256 checksum

### ISSUE THREAD

**Issue by MageJohn** [ziglang/zig#4945]
I'm using Manjaro, which is Arch Linux based and has access to the AUR, so I'm getting Zig by using the [zig-dev-static](https://aur.archlinux.org/packages/zig-dev-static/) package. However I can't use the one at that link directly because it can be out of date within hours. I download the script and manually update the version and sha256 sums in it. Well, actually I have a bash script that uses the JSON API to do it for me. But I had to write that manually. The point is there are a couple of things that would be nice to have which would make packaging easier:

- A consistent link for the latest build. Something like https://ziglang.org/builds/zig-linux-x86_64-latest.tar.xz. It's not possible to get makepkg to somehow use the JSON API.
- A PGP signature for the build. Again, the SHA sum from the JSON API can't be used, but a signature file can be added to the sources array, and will be used to automatically check the file.
- A monotonically increasing version number. Right now the latest version is often registered as a downgrade, because of the commit id. The easiest way is to use git describe --long, or even better follow the recommendation on the Arch Wiki: https://wiki.archlinux.org/index.php/VCS_package_guidelines#The_pkgver()_function. The version number can be updated with a function in the PKGBUILD (only after the sources have been downloaded), so the JSON API can be used. 

These changes would allow installing the latest Zig build directly from the AUR with zero friction. And hopefully they're helpful to more than just AUR users as well!

**Comment by andrewrk** [comment]
The other two are possible but unfortunately the consistent link for the latest build is tricky due to limitations of s3, the (current) backend of ziglang.org. We used to have this but it would be incorrectly cached and ended up causing more problems than it solved.

**Comment by MasterQ32** [comment]
I have written a [`zig-up script`](https://gist.github.com/MasterQ32/1f27081d8a7dd109fc290bd6729d97a8) that will update a locally installed zig version to the latest build from the website, it depends on bash, tar and jq, script is tested and used on an archlinux, but needs adjustments to other systems (and maybe user names *grin*)

As zig has no external dependencies, it should be enough to just dispatch this script (or even better: rewrite it in zig!) to provide the current master build of zig.

**Comment by MageJohn** [comment]
Some sort of zig-up tool would be useful, indeed. Especially if it also helps manage multiple versions of Zig like rustup does. The language is evolving fast, and sometimes introduces incompatibilities that might mean that someone has projects on the same machine requiring different versions of Zig. This may not be the issue to discuss this though.

**Comment by andrewrk** [comment]
> * A consistent link for the latest build.

Sadly impossible due to using S3 for ziglang.org.

> * A PGP signature for the build.

Planned starting with the upcoming 0.9.0 release. I'll keep this issue open to track it.

> * A monotonically increasing version number.

Done since #6509.

> Some sort of zig-up tool would be useful, indeed

There are existing community solutions to this, such as [zigup](https://github.com/marler8997/zigup/). See also  #10204.

**Comment by andrewrk** [comment]
Alright I still want to do this but I'm not going to try to rush it 2 weeks before the 0.9.0 release. Will take my time setting this up and do it as part of the 0.10.0 release cycle.

**Comment by jedisct1** [comment]
Minisign is also an option for signing builds, especially since there's a pure Zig version that can be used for verification.

Minisign signatures are also supported by the Sigstore transparency log, which is an important piece of the puzzle for supply chain security.

**Comment by Techcable** [comment]
This is a problem for me because it makes it tough for users to reliably download as part of a AUR-style build system.

The AUR package for [`zig-dev-bin`](https://aur.archlinux.org/packages/zig-dev-bin) is in a tricky situation:

Generally speaking, AUR packages are expected to either GPG or SHA256 verify the files they are downloading.

Right now now GPG verification is not possible (due to the issue). Ideally the package would auto-updating to the latest hightly build, but that would make hardcoding sha256 checksums impossible.

The sha256 in `version-index.json` is not helpful for my purposes. It just moves the integrity problem into a different file.

Right now the current approach the maintainer makes is to pin against specific nightly versions and use a hardcoded SHA256 hash. He updates it every couple of weeks. This is obviously not a great approach (it requires manual intervention by maintainer, and uses outdated builds), but the alternative would be "unverified" files, relying only on the integrity of the https network connection :(

I have a [patch](https://gist.github.com/Techcable/a697053de49a644d22b438cf4b3db693) to make the aur package auto-update, but I'm not sure the maintainer will accept it for this integrity reason. It might even be against AUR policy.

I know this complaining isn't particularly helpful, but I thought I'd at least bring up my use-case :)

I think someone with access to the build infrastructure would have to be the one who ultimately sets this up.
A
lso it's fine if the GPG key is a dummy value (like "build.bot@ziglang.org",). There is no need to use a personal/organization GPG key, it's just important a signature exists :wink:

**Comment by maxmilton** [comment]
I'm on Arch Linux and ran into the problem of the outdated AUR package as well.

Rather than wait for this, I've created [my own small bash script](https://gist.github.com/maxmilton/dc4a88cb100bda7b0ee42ce582b47bfe) that also verifies the shasum hash and file size. It's based on the gist shared before, thanks for that @MasterQ32 !

**Comment by Techcable** [comment]
> Rather than wait for this, I've created [my own small bash script](https://gist.github.com/maxmilton/dc4a88cb100bda7b0ee42ce582b47bfe) that also verifies the shasum hash and file size. It's based on the gist shared before, thanks for that @MasterQ32 !

Really, this does not solve the underlying security problem. If someone can intercept your HTTPS connection to the binary they can also intercept your connection to `index.json`. It's the exact same domain name and similar url.

You are implicitly trusting that the HTTPS connection to https://ziglang.org/download/index.json is valid.
If someone was able to forge a trusted TLS certificate, someone could hijact your connection and present you with invalperspectiveid (although very difficult to do, it has happened before.

Checking the sha256sum does protect against honest mistakes or download errors (which happen a good deal), but downloading the checksum from the same server as the binary does not provide any integrity checks from a security .

At this point you're probably thinking "this is crazy" and "why would anyone want to intercept *my* HTTPS connection". I agree. I don't personally worry about anyone intercepting my HTTPS connections. 

However, I am under the impression that PGP signing nightly binaries is a security best-practice. It is certainly strongly suggested on the AUR. I think we should move in that direction 

This is what Rust does (even for their nightly builds). Here is a link to the latest nightly binary and ts corresponding GPG signature:
1.https://static.rust-lang.org/dist/rust-nightly-aarch64-unknown-linux-musl.tar.gz
2. https://static.rust-lang.org/dist/rust-nightly-aarch64-unknown-linux-musl.tar.gz.asc (small text file, safe to click)

These are both signed by the [rust signing key](https://static.rust-lang.org/rust-key.gpg.ascii)

**Comment by maxmilton** [comment]
@Techcable that makes sense to me, especially given I fetch the shasum dynamically. Thank you for explaining your thinking!

A (dynamically fetched) checksum to validate the downloaded file integrity is nice, but I agree with you, tightly controlled PGP key/s sounds like a very good idea for security :+1:

**Comment by andrewrk** [comment]
@jedisct1 I'm looking into setting this up. Do you recommend minisign? If so, I'll move forward with it.

**Comment by jedisct1** [comment]
Minisign has [quite a few implementations](https://github.com/jedisct1/minisign#additional-tools-libraries-and-implementations) including in Zig (for verification only; I can quickly add signing).

It's also supported by Google's transparency log https://docs.sigstore.dev/rekor/sign-upload/

An alternative would be to simply use OpenSSH that can now be used to sign arbitrary files: https://www.agwa.name/blog/post/ssh_signatures

**Comment by andrewrk** [comment]
How can I make minisign not prompt for an (empty) password?

**Comment by andrewrk** [comment]
Done in https://github.com/ziglang/www.ziglang.org/commit/1164e10206975b3940362b58419a53c2149730a2. Master branch signatures will arrive with the next run of the build-tarballs script.

**Comment by jayschwa** [comment]
@andrewrk, `*.minisig` files on the download page are currently returning "403 Access Denied".

**Comment by jedisct1** [comment]
> How can I make minisign not prompt for an (empty) password?
> 
> Edit: looks like `echo | minisign ...` works. But that's going to be problematic in a bash script, if the `minisign` binary returns an error code.

`echo | minisign` or `minisign < passwd.txt`, but I can add a flag to not prompt for a password. I think the Rust implementation has that feature, so let's do it in the C version as well.

**Comment by andrewrk** [comment]
The first run of tarballs script failed because I missed a line; fixed in https://github.com/ziglang/www.ziglang.org/commit/015540fbb02addec80da6ce98dde395eccb2ca41. Fix should arrive on the website in about 9 hours.

**Comment by jedisct1** [comment]
`minisign 0.11` is going to be released now.

The `-W` flag skips key encryption; however you'll need to create a new key pair with that flag set.

**Comment by mitchellh** [comment]
I just want to propose that we keep the SHA hashes AND sign the archives. The reason being is that various package managers (Nix, Arch) require a SHA256 sum ahead of time to download _anything_. So, with my Nix Zig packages, I'm now forced to download the releases to compute the sum, store that in the package descriptor, AND verify signatures.

My usual approach I've taken with this for my software is to ship a SHA256SUMs file alongside all of my releases, and PGP sign that as well. So you end up with basically the following:

  - `zig-1234.tar.xz`
  - `zig-1234.tar.xz.minisig`
  - `zig-1234-SHA256SUMS` (probably the sums for every platform here)
  - `zig-1234-SHA256SUMS.minisig`

For Zig, perhaps the most helpful would be to provide `index.json` and `index.json.minisg`, and have the checksums be in the JSON file. Package managers have to get a SHA256SUMs-esque file from somewhere anyways (or compute it themselves), and downloading the one JSON file + one small sig is a lot easier to setup the metadata in a package. 

The package itself could download the binary and the binary signature and verify those as part of install. 

**TLDR:** Re-add the checksums, sign the `index.json` file, and some package management pain goes away.

Not something that's terribly important to me but figured I'd share this experience report.

**Comment by nektro** [comment]
is there any prior art in anyone making a web server that generates these on the fly?

**Comment by MasterQ32** [comment]
> is there any prior art in anyone making a web server that generates these on the fly?

That would be a horrible waste of resources, considering the tradeoff {storing 16 byte} vs {computing 16 byte from 100 MB each request}

**Comment by nektro** [comment]
`zig-windows-x86_64.tar.xz` is almost double the size of the next largest tarball at 71MB and the sha256 can be calculated in only .03s

**Comment by andrewrk** [comment]
Thanks for the feedback @mitchellh. I agree with your conclusions and I will make it so.

**Comment by andrewrk** [comment]
Done in https://github.com/ziglang/www.ziglang.org/commit/af9d69d5f284089d4c9a16acf4e926e9da5e90b1. Checksums for master branch will arrive with the next CI run. The signature of the index.json file itself is located at `https://ziglang.org/builds/zig-$VERSION-index.json.minisig`. Unfortunately, making it `https://ziglang.org/download/index.json.minisig` would be a race condition.

**Comment by mitchellh** [comment]
Thank you, that made things very easy. I now verify the signature of the index.json then trust the hashes in the manifest: https://github.com/mitchellh/zig-overlay/commit/dccddf644a02e9afe4b5c289dae97aa06284f245
